source/common/common/utility.h

```diff
@@ -636,9 +636,66 @@ class WelfordStandardDeviation {
   double m2_{0};
 };
 
-template <class Value> struct TrieEntry {
+// A TrieEntry aims to be a good balance of performant and
+// space-efficient, by allocating a vector the size of the range of children
+// the node contains. This should be good for most use-cases.
+//
+// For example, a node with children 'a' and 'z' will contain a vector of
+// size 26, containing two values and 24 nulls. A node with only one
+// child will contain a vector of size 1. A node with no children will
+// contain an empty vector.
+//
+// Compared to allocating 256 entries for every node, this makes insertions
+// a little bit inefficient (especially insertions in reverse order), but
+// trie lookups remain O(length-of-longest-matching-prefix) with just a
+// couple of very cheap operations extra per step.
+//
+// By size, having 256 entries for every node makes each node's overhead
+// (excluding values) consume 8KB; even a trie containing only a single
+// prefix "foobar" consumes 56KB.
+// Using ranged vectors like this makes a single prefix "foobar" consume
+// less than 20 bytes per node, for a total of 0.14KB.
```

Comment:
Can you describe in the comment here what this type expects of Value? I see below we have a line `return {}` from a function that returns a Value. I'm assuming what that means is that `Value` must be empty-constructable.

But I see in the previous state of this code in that the function returned nullptr, which makes me think that Value must be a pointer (or unique_ptr or shared_ptr).

Reply:
Added in TrieLookupTable comment since TrieEntry isn't even really the public API.

My change to {} rather than nullptr was to reduce the strictness of the constraint - there's no reason it has to be a smart pointer, but it does have to be empty-initializable and comparable. I don't think it really needed a comment saying so, because if you were to try to put any incompatible value in it it won't compile.

Reply:
What you put there is fine, though I'd prefer more helpful comments than forcing the user to discover what's possible by writing code and then seeing if it compiles :)

But if TrieEntry is no supposed to be accessed by users, can we make it a private nested class of TrieLookupTable?